gt/main.c

```diff
@@ -119,10 +119,199 @@ lookup_policy_decision(void *ip_hdr, uint8_t ip_version,
 	return 0;
 }
 
+static void
+fill_eth_hdr_reverse(struct ether_hdr *eth_hdr,
+	struct ipip_tunnel_info *tunnel)
+{
+	ether_addr_copy(&tunnel->source_mac, &eth_hdr->d_addr);
+	ether_addr_copy(&tunnel->nexthop_mac, &eth_hdr->s_addr);
+	eth_hdr->ether_type = rte_cpu_to_be_16(tunnel->flow.proto);
+}
+
+static uint16_t
+get_psd_sum(void *l3_hdr, uint16_t ethertype, uint64_t ol_flags)
+{
+	if (ethertype == ETHER_TYPE_IPv4)
+		return rte_ipv4_phdr_cksum(l3_hdr, ol_flags);
+	else
+		return rte_ipv6_phdr_cksum(l3_hdr, ol_flags);
+}
+
+static int
+fill_notify_pkt(unsigned int socket,
+	struct rte_mbuf *notify_pkt, struct ggu_policy *policy,
+	struct ipip_tunnel_info *tunnel, struct gt_config *gt_conf)
+{
+	uint8_t *data;
+	uint16_t ethertype = tunnel->flow.proto;
+	struct ether_hdr *notify_eth;
+	struct ipv4_hdr *notify_ipv4;
+	struct ipv6_hdr *notify_ipv6;
+	struct udp_hdr *notify_udp;
+	struct ggu_common_hdr *notify_ggu;
+
+	notify_pkt = rte_pktmbuf_alloc(
+		gt_conf->net->gatekeeper_pktmbuf_pool[socket]);
+	if (notify_pkt == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gt: failed to allcate notification packet!");
+		return -1;
+	}
+
+	if (ethertype == ETHER_TYPE_IPv4) {
+		notify_eth = (struct ether_hdr *)rte_pktmbuf_prepend(notify_pkt,
+			sizeof(struct ether_hdr) + sizeof(struct ipv4_hdr) +
+			sizeof(struct udp_hdr) + sizeof(struct ggu_common_hdr));
+		notify_ipv4 = (struct ipv4_hdr *)&notify_eth[1];
+		notify_udp = (struct udp_hdr *)&notify_ipv4[1];
+		notify_ggu = (struct ggu_common_hdr *)&notify_udp[1];
+	} else if (ethertype == ETHER_TYPE_IPv6) {
+		notify_eth = (struct ether_hdr *)rte_pktmbuf_prepend(notify_pkt,
```

Comment:
Given that the packet pointed by `notify_pkt` is empty, and headroom is smaller than the data that the packet can carry, shouldn't we use `rte_pktmbuf_append()` instead of `rte_pktmbuf_prepend()` in both cases above?